Convert IPAM to use admiral resource syncer

tpantelis opened the pull request:
Fixes https://github.com/submariner-io/submariner/issues/375

tpantelis committed:
Convert IPAM to use admiral resource syncer

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

mangelajo commented:
Hi @tpantelis although I see the value of using common code in terms of maintenance I'm stressed out by the risks of the 0.9 release, I have preliminary assigned this one to 0.10, but feel free to bring it up in team meeting for rescheduling.

On the other hand IPAM/globalnet is something we have discarded for full support until we make globalnet v2 in 0.10, so may be it's simply not that risky

skitt commented:
> Hi @tpantelis although I see the value of using common code in terms of maintenance I'm stressed out by the risks of the 0.9 release, I have preliminary assigned this one to 0.10, but feel free to bring it up in team meeting for rescheduling.
> 
> On the other hand IPAM/globalnet is something we have discarded for full support until we make globalnet v2 in 0.10, so may be it's simply not that risky

It seems to me that this would simplify support by only requiring us to fix bugs in one set of code, and ensuring that any bug fixes are applied everywhere.

sridhargaddam commented:
This PR is very intrusive and needs some dedicated time to ensure that there are no regressions. I will take a look when I get some bandwidth.

mangelajo commented:
Based on @sridhargaddam  commends and mine, I think it's better if we wait for 0.10.

I don't want to introduce errors on globalnet at this point, where we should be using our full bandwidth on the 0.9 features we planned.

mangelajo reviewed (changes requested):
I vote we wait for 0.10

There are no plans to touch the globalnet code, so the code should not age badly.

tpantelis commented:
I didn't mean to create a ruckus - sorry 😄 This is fine for 0.10. This is something that's been on my list and I had started before the priority shift. Note that we've seen intermittent test failures due to the structure of the prior code - converting to the admiral APIs fixes that.  The main reason I wanted to at least get the patch done.

tpantelis commented:
> This PR is very intrusive and needs some dedicated time to ensure that there are no regressions. I will take a look when I get some bandwidth.

Understood. Regression was the reason I created the unit tests prior to refactoring. The pertinent code paths are covered by the tests.

mangelajo commented:
> I didn't mean to create a ruckus - sorry  This is fine for 0.10. This is something that's been on my list and I had started before the priority shift. Note that we've seen intermittent test failures due to the structure of the prior code - converting to the admiral APIs fixes that. The main reason I wanted to at least get the patch done.

Thank you, it makes a lot of sense ':D

tpantelis commented:
bump

sridhargaddam commented:
bump

tpantelis commented:
bump

mangelajo commented on pkg/globalnet/controllers/ipam/gatewaymonitor.go:
```diff
@@ -16,63 +16,74 @@ limitations under the License.
+			Name:         "IPAM GatewayMonitor",
+			ResourceType: &v1.Endpoint{},
+			Handler: watcher.EventHandlerFuncs{
+				OnCreateFunc: gatewayMonitor.handleCreatedOrUpdatedEndpoint,
```
What do you think about moving right away to the https://github.com/submariner-io/submariner/blob/devel/pkg/event/handler.go interface instead.

The event handler also provides updates on the nodes. As well as transitions to gateways or non-gateways.

@sridhargaddam  @tpantelis ^ ?

tpantelis replied:
yeah we can look into using that interface but I'd rather do that in another PR. We need to move this forward to proceed with GN 2.0 enhancements.

mangelajo replied:
@tpantelis that sounds good this PR already implements a good deal of improvements and testing. It should be easier to move from this one if we finally see that as reasonable.

mangelajo commented on pkg/globalnet/controllers/ipam/gatewaymonitor.go:
```diff
@@ -124,197 +127,104 @@ func (gm *GatewayMonitor) Stop() {
-	return true
-}
+	// If the endpoint hostname matches with our hostname, it implies we are on gateway node
+	if endpoint.Spec.Hostname == hostname {
```
This detection is also provided by the event interface. (Transition to gateway and transition to non gateway)

mangelajo reviewed (commented):
I made one question.

mangelajo commented:
as @tpantelis commented wireguard+globalnet failed, but we need to determine if it's really related to the patch:

https://github.com/submariner-io/submariner/pull/1196/checks?check_run_id=2556449601

The error here was on gateway failover.

I will rerun

For the flake finder we see ~15 days ago some failures with libreswan + globalnet or wireguard+globalnet

https://github.com/submariner-io/submariner/actions/workflows/flake_finder.yml

In that case it was also redundency related, but in a different point:
```
[Fail] [redundancy] Route Agent restart tests when a route agent pod running on a gateway node is restarted [It] should start a new route agent pod and be able to connect from another cluster 
```

sridhargaddam commented on pkg/globalnet/controllers/ipam/gatewaymonitor.go:
```diff
@@ -124,197 +127,104 @@ func (gm *GatewayMonitor) Stop() {
-			klog.Errorf("Could not convert object %v to an Endpoint", obj)
-			return
+		gm.syncMutex.Lock()
+		if !gm.isGatewayNode {
```
This part of code gets simplified if we move to Event handler interface, but that can be done in a separate PR.

mangelajo commented:
It failed again with wireguard on failover: https://github.com/submariner-io/submariner/pull/1196/checks?check_run_id=2564983655#step:3:2912

sridhargaddam commented:
The code looks good to me, but we have to investigate why it's failing with Wireguard driver alone.

sridhargaddam commented on pkg/globalnet/controllers/ipam/gatewaymonitor.go:
```diff
@@ -124,197 +127,104 @@ func (gm *GatewayMonitor) Stop() {
-		endpoints, err := gm.submarinerClientSet.SubmarinerV1().Endpoints(ns).List(context.TODO(), metav1.ListOptions{})
+		overlap, err := cidr.IsOverlapping(endpoint.Spec.Subnets, gm.ipamSpec.GlobalCIDR[0])
 		if err != nil {
-			gm.endpointWorkqueue.Forget(obj)
-			return fmt.Errorf("error retrieving submariner endpoint list %v", err)
+			// Ideally this case will never hit, as the subnets are valid CIDRs
+			klog.Warningf("unable to validate overlapping Service CIDR: %s", err)
 		}
 
-		for _, endpoint := range endpoints.Items {
-			if endpoint.Spec.ClusterID != gm.clusterID {
-				for _, remoteSubnet := range endpoint.Spec.Subnets {
-					MarkRemoteClusterTraffic(gm.ipt, remoteSubnet, AddRules)
-				}
-			}
-		}
```
@tpantelis I was debugging the issue with Wireguard driver and found the problem. The following code is accidentally removed in the new implementation.

I was wondering why it was working with Libreswan Cabledriver and noticed that there is a difference of behavior between the two. In Libreswan driver, when the Gateway pod is restarted, it was not deleting the local Endpoint but with Wireguard driver, it was deleting the local endpoint. Anyways, IMHO I feel we should fix the code in Globalnet.

tpantelis replied:
Thanks for looking into this. I remember this code now - it didn't seem necessary since we do the same thing above on L173 when we get notified of the remote Endpoint. So I'm not clear on why we would also do this for a local Endpoint, ie it seems redundant.

tpantelis replied:
Fixed with https://github.com/submariner-io/submariner/pull/1196/commits/63cc50ef87f7a04d3b8e3ab47c7fc7be7ed8cd66.

sridhargaddam replied:
Thanks for fixing it @tpantelis

sridhargaddam commented on pkg/globalnet/controllers/ipam/gatewaymonitor.go:
```diff
@@ -124,197 +127,104 @@ func (gm *GatewayMonitor) Stop() {
+		return false
+	}
 
+	hostname, err := os.Hostname()
```
Here is the missing code : https://github.com/submariner-io/submariner/blob/7c57fa2974dab3c57b9e870ded0e7c61ea2649f2/pkg/globalnet/controllers/ipam/gatewaymonitor.go#L183

tpantelis committed:
Add IP table rules for remote subnets on local EP add

The IP table rules for remote Endpoints are cleared when the local
Endpoint is deleted. If the gateway was simply restarted on the same node
then we need to re-add the rules when the new local Endpoint is created.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

sridhargaddam commented:
@mangelajo can you please re-review this PR. Its now passing CI as well.
